usr/lib/ica_s390_stdll/ica_specific.c

```diff
@@ -3124,6 +3125,364 @@ static CK_RV os_specific_rsa_decrypt(STDLL_TokData_t *tokdata,
+    UNUSED(tokdata);
+    UNUSED(sess);
+
+    context = (AES_GCM_CONTEXT *)ctx->context;
+    gcm_ctx = (kma_ctx *)context->ulClen;
+    aes_gcm_param = (CK_GCM_PARAMS *)ctx->mech.pParameter;
+    auth_data = (CK_BYTE *)aes_gcm_param->pAAD;
+    auth_data_len = aes_gcm_param->ulAADLen;
+    tag_data = out_data + in_data_len;
+    tag_data_len = (aes_gcm_param->ulTagBits + 7) / 8;
+
+    /* perform one-part encryption/decryption */
+    rc = ica_aes_gcm_kma_update(in_data, out_data,
+                                encrypt ? in_data_len : in_data_len - tag_data_len,
+                                auth_data, auth_data_len,
+                                1, 1, gcm_ctx);
+    if (rc != 0) {
+        TRACE_ERROR("ica_aes_gcm_kma_update failed, rc=%ld.\n", rc);
+        (*out_data_len) = 0;
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    }
+
+    if (encrypt) {
+        /* encrypt: provide authentication tag and append it to out_data */
+        rc = ica_aes_gcm_kma_get_tag(tag_data, tag_data_len, gcm_ctx);
+        if (rc == 0) {
+            (*out_data_len) = in_data_len + tag_data_len;
+        } else {
+            TRACE_ERROR("ica_aes_gcm_kma_get_tag failed, rc=%ld.\n", rc);
+            (*out_data_len) = 0;
+            rc = CKR_FUNCTION_FAILED;
+            goto done;
+        }
+    } else {
+        /* decrypt: verify tag */
+        tag_data = in_data + in_data_len - tag_data_len;
+        rc = ica_aes_gcm_kma_verify_tag(tag_data, tag_data_len, gcm_ctx);
+        if (rc == 0) {
+            (*out_data_len) = in_data_len - tag_data_len;
+        } else {
+            TRACE_ERROR("ica_aes_gcm_kma_verify_tag failed, rc=%ld.\n", rc);
+            (*out_data_len) = 0;
+            rc = CKR_FUNCTION_FAILED;
+            goto done;
+        }
+    }
+
+done:
+
+    ica_aes_gcm_kma_ctx_free(gcm_ctx);
+    context->ulClen = (CK_ULONG)0;
+
+    return rc;
+}
+
+CK_RV new_gcm_specific_aes_gcm_update(STDLL_TokData_t *tokdata, SESSION *sess,
+                                      ENCR_DECR_CONTEXT *ctx, CK_BYTE *in_data,
+                                      CK_ULONG in_data_len, CK_BYTE *out_data,
+                                      CK_ULONG *out_data_len, CK_BYTE encrypt)
+{
+    AES_GCM_CONTEXT *context = NULL;
+    CK_GCM_PARAMS *aes_gcm_param = NULL;
+    kma_ctx *gcm_ctx = NULL;
+    CK_BYTE *auth_data, *buffer = NULL;
+    CK_ULONG total, tag_data_len, remain, auth_data_len, out_len;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+    UNUSED(sess);
+    UNUSED(encrypt); // direction already set in gcm ctx
+
+    context = (AES_GCM_CONTEXT *)ctx->context;
+    gcm_ctx = (kma_ctx *)context->ulClen;
+    aes_gcm_param = (CK_GCM_PARAMS *)ctx->mech.pParameter;
+    auth_data = (CK_BYTE *)aes_gcm_param->pAAD;
+    auth_data_len = aes_gcm_param->ulAADLen;
+    tag_data_len = (aes_gcm_param->ulTagBits + 7) / 8;
+    total = (context->len + in_data_len);
+
+    /* if there isn't enough data to make a block, just save it */
+    if (encrypt) {
```

Comment:
above you make `encrypt` UNUSED (even with a comment), but here you are still using it ?

Reply:
Argh, the final routine ... yes, removed it.

Reply:
Hm... in line 3270 in this function ist still marked as UNUSED.

Reply:
Hmm ... I don't have it in my local version of ica_specific.c ...

Reply:
its still there
![image](https://github.com/opencryptoki/opencryptoki/assets/28868064/61b6f4ce-95c2-4aec-98e6-fa6513e8cf21)